src/backend/access/transam/xlog.c

```diff
@@ -7031,7 +7043,15 @@ StartupXLOG(void)
 		EndRecPtr = ControlFile->checkPointCopy.redo;
 
 		memcpy(&checkPoint, &ControlFile->checkPointCopy, sizeof(CheckPoint));
-		wasShutdown = true;
+		// When normal (primary) Neon compute node is started, we assume that is started after normal shutdown and
+		// no recovery is needed.
+		// When read-only replica is started, we need to obtain information about running xacts, so wasShutdown is set to false.
+		// When snapshot read-only node is started, we can treat all active transactions as aborted so once again
+		// assume that we restart after normal shutdown.
+		wasShutdown = StandbyModeRequested
+			&& PrimaryConnInfo != NULL && *PrimaryConnInfo != '\0'
+			&& IsPrimaryAlive()
+			? false : true;
```

Comment:
Some comment kibitzing, and I think the condition is more readable this way, so that it matches the flow of the comment:

```suggestion
		// When primary Neon compute node is started, we pretend that it started after a clean shutdown and
		// no recovery is needed. We don't need to do WAL replay, the page server does that on a page-by-page basis.
		// When a read-only replica is started, PostgreSQL normally waits for a shutdown checkpoint or running-xacts
		// record before enabling hot standby, to establish which transactions are still running in the primary,
		// and might still commit later. But if we know that the primary is not running - because the control plane
		// says so - we can skip that. That avoids having to wait indefinitely if the primary is not running. This is
		// particularly important for Neon because we don't start recovery from a checkpoint record, so there's
		// no guarantee on when we'll see the next checkpoint or running-xacts record, if ever. so if we know the primary is
		// not currently running, also set wasShutdown to 'true'.
		if (StandbyModeRequested &&
			PrimaryConnInfo != NULL && *PrimaryConnInfo != '\0')
		{
			if (!IsPrimaryAlive())
				wasShutdown = true;
			else
				wasShutdown = false;
		}
		else
			wasShutdown = true;
```

Reply:
Why the condition on PrimaryConnInfo though? Is that to set 'wasShutdown = true' on a read-only replica that's "anchored" to a specific LSN? Lets explain that case explicitly in the comment too.

PrimaryConnInfo can be changed on the fly. Is it possible that it's empty when the replica is started, but it's set later? Maybe not possible currently, but I'm worried that this will bite us later. Is there a more direct way of checking for an "LSN-anchored" replica?